`DATE_DIFF()` with just two args fails with NPE

### Description

`row date_diff("year", "2021-12-12")` and `row date_diff("2021-12-12", "2022-12-12")` will both NPE.

```
Caused by: java.lang.NullPointerException
	at java.base/java.util.Objects.requireNonNull(Objects.java:233)
	at java.base/java.util.ImmutableCollections.listFromTrustedArray(ImmutableCollections.java:215)
	at java.base/java.util.List.of(List.java:972)
	at org.elasticsearch.xpack.esql.expression.function.scalar.date.DateDiff.<init>(DateDiff.java:165)
	at org.elasticsearch.xpack.ql.expression.function.FunctionRegistry.lambda$def$8(FunctionRegistry.java:262)
	at org.elasticsearch.xpack.ql.expression.function.FunctionRegistry.lambda$def$3(FunctionRegistry.java:172)
	at org.elasticsearch.xpack.ql.expression.function.FunctionResolutionStrategy.buildResolved(FunctionResolutionStrategy.java:27)
	at org.elasticsearch.xpack.ql.expression.function.UnresolvedFunction.buildResolved(UnresolvedFunction.java:81)
	at org.elasticsearch.xpack.esql.analysis.Analyzer$ResolveFunctions.resolveFunction(Analyzer.java:757)
	at org.elasticsearch.xpack.esql.analysis.Analyzer$ResolveFunctions.lambda$rule$0(Analyzer.java:739)
	at org.elasticsearch.xpack.ql.tree.Node.lambda$transformDown$9(Node.java:185)
	at org.elasticsearch.xpack.ql.tree.Node.transformDown(Node.java:176)
	at org.elasticsearch.xpack.ql.tree.Node.lambda$transformDown$8(Node.java:179)
	at org.elasticsearch.xpack.ql.tree.Node.transformChildren(Node.java:211)
	at org.elasticsearch.xpack.ql.tree.Node.transformDown(Node.java:179)
	at org.elasticsearch.xpack.ql.tree.Node.transformDown(Node.java:185)
	at org.elasticsearch.xpack.ql.plan.QueryPlan.lambda$transformExpressionsOnly$3(QueryPlan.java:87)
	at org.elasticsearch.xpack.ql.plan.QueryPlan.doTransformExpression(QueryPlan.java:113)
	at org.elasticsearch.xpack.ql.plan.QueryPlan.doTransformExpression(QueryPlan.java:125)
	at org.elasticsearch.xpack.ql.plan.QueryPlan.lambda$transformExpressionsOnly$4(QueryPlan.java:87)
	at org.elasticsearch.xpack.ql.tree.NodeInfo.lambda$transform$0(NodeInfo.java:56)
	at org.elasticsearch.xpack.ql.tree.NodeInfo$2.innerTransform(NodeInfo.java:92)
	at org.elasticsearch.xpack.ql.tree.NodeInfo.transform(NodeInfo.java:60)
	at org.elasticsearch.xpack.ql.tree.Node.transformNodeProps(Node.java:260)
	at org.elasticsearch.xpack.ql.tree.Node.transformPropertiesOnly(Node.java:240)
	at org.elasticsearch.xpack.ql.plan.QueryPlan.transformExpressionsOnly(QueryPlan.java:87)
	at org.elasticsearch.xpack.esql.analysis.Analyzer$ResolveFunctions.rule(Analyzer.java:737)
	at org.elasticsearch.xpack.esql.analysis.Analyzer$ResolveFunctions.rule(Analyzer.java:733)
	at org.elasticsearch.xpack.ql.analyzer.AnalyzerRules$ParameterizedAnalyzerRule.lambda$apply$0(AnalyzerRules.java:103)
	at org.elasticsearch.xpack.ql.tree.Node.lambda$transformUp$11(Node.java:198)
	at org.elasticsearch.xpack.ql.tree.Node.transformUp(Node.java:192)
	at org.elasticsearch.xpack.ql.tree.Node.transformUp(Node.java:198)
	at org.elasticsearch.xpack.ql.analyzer.AnalyzerRules$ParameterizedAnalyzerRule.apply(AnalyzerRules.java:103)
	at org.elasticsearch.xpack.ql.analyzer.AnalyzerRules$ParameterizedAnalyzerRule.apply(AnalyzerRules.java:95)
	at org.elasticsearch.xpack.ql.rule.ParameterizedRuleExecutor.lambda$transform$0(ParameterizedRuleExecutor.java:29)
	at org.elasticsearch.xpack.ql.rule.RuleExecutor$Transformation.<init>(RuleExecutor.java:88)
	at org.elasticsearch.xpack.ql.rule.RuleExecutor.executeWithInfo(RuleExecutor.java:167)
	at org.elasticsearch.xpack.ql.rule.RuleExecutor.execute(RuleExecutor.java:136)
	at org.elasticsearch.xpack.esql.analysis.Analyzer.analyze(Analyzer.java:142)
	at org.elasticsearch.xpack.esql.session.EsqlSession.lambda$analyzedPlan$2(EsqlSession.java:157)
	at org.elasticsearch.xpack.esql.session.EsqlSession.lambda$preAnalyze$8(EsqlSession.java:196)
	at org.elasticsearch.server@8.15.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$ResponseWrappingActionListener.onResponse(ActionListenerImplementations.java:245)
	... 22 more
"

```

--- a/docs/changelog/108834.yaml
+++ b/docs/changelog/108834.yaml
@@ -1,0 +1,6 @@
+pr: 108834
+summary: "[ESQL] Mark `date_diff` as requiring all three arguments"
+area: ES|QL
+type: bug
+issues:
+ - 108383

--- a/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/date/DateDiff.java
+++ b/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/date/DateDiff.java
@@ -48,7 +48,7 @@ import static org.elasticsearch.xpack.ql.type.DataTypeConverter.safeToInt;
  * in multiples of the unit specified in the first argument.
  * If the second argument (start) is greater than the third argument (end), then negative values are returned.
  */
-public class DateDiff extends EsqlScalarFunction implements OptionalArgument {
+public class DateDiff extends EsqlScalarFunction {
 
     public static final ZoneId UTC = ZoneId.of("Z");
 

--- a/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/date/DateDiff.java
+++ b/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/date/DateDiff.java
@@ -17,7 +17,6 @@ import org.elasticsearch.xpack.esql.expression.function.Param;
 import org.elasticsearch.xpack.esql.expression.function.scalar.EsqlScalarFunction;
 import org.elasticsearch.xpack.ql.InvalidArgumentException;
 import org.elasticsearch.xpack.ql.expression.Expression;
-import org.elasticsearch.xpack.ql.expression.function.OptionalArgument;
 import org.elasticsearch.xpack.ql.tree.NodeInfo;
 import org.elasticsearch.xpack.ql.tree.Source;
 import org.elasticsearch.xpack.ql.type.DataType;

